### PR TITLE
feat: tls jobs comparing http.sys win vs kestrel linux

### PR DESCRIPTION
**Pull request by DeagleGross**
The goal here is to compare specific TLS related scenarios using "Windows Http.Sys" and "Linux Kestrel". I've done testing via crank against `aspnet-perf-win` and `aspnet-perf-lin`.

TLS handshake is established once per connection, therefore client is passing `Connection: close` header every request, meaning that connection will be re-established every request including the TLS handshake. To make sure that scenario works properly, I have added the app args like `statsEnabled`, which write to the logs information about how many connections were established, and how many certs were fetched and visible from the c# code. Example (numbers are not accurate due to concurrency):
image

## TLS handshakes
The first benchmark is to measure only TLS handshakes scenario: `tls-handshakes-httpsys` and `tls-handshakes-kestrel`.
**win http.sys** RPS = 2,807
```csharp
builder.WebHost.UseHttpSys();

app.MapGet("/hello-world", () =>
{
    return Results.Ok("Hello World!");
});
```

**lin kestrel** RPS = 2,000
```csharp
builder.WebHost.UseKestrel(options =>
    ...
    serverOptions.Listen(endpoint, listenOptions =>
    {
        // [SuppressMessage("Microsoft.Security", "CSCAN0220.DefaultPasswordContexts", Justification="Benchmark code, not a secret")]
        listenOptions.UseHttps("testCert.pfx", "testPassword");
    ...
```

## mutual TLS handshakes
In mutual TLS handshakes scenario both client and server send their certificates to each other.

**win http.sys** RPS = 900
in case of HTTP.SYS the only usable API is [ConnectionInfo.GetClientCertificateAsync](https://learn.microsoft.com/en-us/dotnet/api/microsoft.aspnetcore.http.connectioninfo.getclientcertificateasync?view=aspnetcore-9.0), which forces to get a client certificate from the connection (that is what we want to let devs write a business logic against the certificate).

In case of Http.Sys certificate validation is basically a middleware, where client cert is fetched and there is no-op call to next middleware in the pipeline.
```csharp
app.Use(async (context, next) => {
    var clientCert = context.Connection.ClientCertificate;
    if (clientCert is null)
    {
        clientCert = await context.Connection.GetClientCertificateAsync(CancellationToken.None);
        fetchedCertsCounter++;
    }

    // we have a client cert here, and lets imagine we do the validation here
    // if (clientCert.Thumbprint != "1234567890") throw new NotImplementedException();

    await next();
});
```

**lin kestrel** RPS = 800
Setup is done by adding requirement on having a client certificate + invoking a callback, which will let devs write their business logic against the certificate.
```csharp
listenOptions.UseHttps("testCert.pfx", "testPassword", options =>
{
    if (mTlsEnabled)
    {
        options.ClientCertificateMode = ClientCertificateMode.RequireCertificate;
        options.ClientCertificateValidation = AllowAnyCertificateValidationWithLogging;
    }
});
```

## TLS renegotiation
This is a scenario, where client can provide a certificate, but initially it is not being used. However later on the opened connection server initiates a TLS handshake by asking client to provide a client certificate.

**win httpsys** RPS = 900
the same scenario as in mTLS because http.sys can't force a client certificate in required mode (at least I am not aware of)

**lin kestrel** RPS = 833
Setup is done by setting `DelayCertificate` mode instead of `RequireCertificate`.
```
options.ClientCertificateMode = ClientCertificateMode.DelayCertificate
```
Moreover, to force reading certificate, I am adding a middleware which will force loading the client certificate:
```
    app.Use(async (context, next) =>
    {
        var clientCert = context.Connection.ClientCertificate;
        if (clientCert is null)
        {
            clientCert = await context.Connection.GetClientCertificateAsync();
        ...
```
logs of the run confirm that certificate is fetched on the middleware only, and only then passed into client certificate validation callback

### CODE REVIEW

File: src/BenchmarksApps/TLS/Kestrel/Program.cs [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+
+app.MapGet("/hello-world", () =>
+{
+    return Results.Ok("Hello World!");
+});
+
+await app.StartAsync();
+
+Console.WriteLine("Application Info:");
+if (mTlsEnabled) Console.WriteLine($"\tmTLS is enabled (client cert is required)");
```
Comment (BrennanConroy) [inline src/BenchmarksApps/TLS/Kestrel/Program.cs:101]: Braces and newlines preferred.

**Review by BrennanConroy — APPROVED** [review]

File: src/BenchmarksApps/TLS/Kestrel/Program.cs (hunk 1 above)
Comment (DeagleGross) [reply src/BenchmarksApps/TLS/Kestrel/Program.cs]: sure, added

File: src/BenchmarksApps/TLS/HttpSys/Program.cs [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+    // this is an http.sys middleware to get a cert
+    Console.WriteLine("Registered client cert validation middleware");
+
+    app.Use(async (context, next) => {
+        var clientCert = context.Connection.ClientCertificate;
+        if (clientCert is null)
+        {
+            if (writeCertValidationEventsToConsole) Console.WriteLine($"No client certificate provided. Fetching for connection {context.Connection.Id}");
+            clientCert = await context.Connection.GetClientCertificateAsync(CancellationToken.None);
+            fetchedCertsCounter++;
```
[the remark this answers is not in the document]
Comment (DeagleGross) [reply src/BenchmarksApps/TLS/HttpSys/Program.cs]: refactored as in your suggestion

File: src/BenchmarksApps/TLS/HttpSys/Program.cs [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+
+#pragma warning disable CA1416 // Can be launched only on Windows (HttpSys)
+builder.WebHost.UseHttpSys(options =>
+{
+    // meaning client can send a certificate, but it can be explicitly requested by server as well (renegotiation)
+    options.ClientCertificateMethod = ClientCertificateMethod.AllowRenegotation;
+});
+#pragma warning restore CA1416 // Can be launched only on Windows (HttpSys)
+
+var config = new ConfigurationBuilder()
```
[the remark this answers is not in the document]
Comment (DeagleGross) [reply src/BenchmarksApps/TLS/HttpSys/Program.cs]: thanks for sharing, I removed building configuration explicitly!

**Review by BrennanConroy — APPROVED** [review]

**Comment by sebastienros** [comment]
I don't remember if adding a pfx won't trigger a creds scan issue :/

**Pull request merged**